Protocol V4 enum support (JAVA-783).

adutra opened the pull request:
This PR basically updates the code and introduces a new `ProtocolVersion` enum constant, namely `V4`, and updates the code in places where a test for the current protocol was made.

A few general remarks:
1. Jorge spotted a problem with C\* 2.1 that affects clients in case of a protocol error response, see CASSANDRA-9451 and class `Connection`;
2. `TupleCodec` and `UserTypeCodec` now are sensitive to protocol versions; this change is not strictly required as of now but I thought better make the change now instead of leaving hard-coded V3 versions in the code;
3. Current tests pass against a 2.1 cluster; testing with C\* 2.2 is obviously not possible yet as v4 is not fully implemented yet;
4. Protocol negotiation with Scassandra tests will fail with strange errors when introducing v4; fixing the protocol version to v2 at cluster level solves the problem (I did not have time to investigate further yet).

Commit:
Protocol V4 enum support (JAVA-783).

adutra commented:
@tolbertam good point, I updated the docs to match the spec files, thanks!

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ArrayBackedRow.java:
```diff
@@ -25,6 +26,13 @@
+     * Note that starting from Cassandra 2.2 built-in functions are declared in
+     * the system keyspace, so the function name is prefixed with "system.".
+     */
+    private static final Pattern TOKEN_COLUMN_NAME = Pattern.compile("(system\\.)?token(.*)");
```
This is due to the fact that now built-in function names are returned with the system prefix.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/ProtocolVersion.java:
```diff
@@ -30,10 +30,12 @@
 
     /**
      * The most recent protocol version supported by the driver.
+     * TODO change to V4 when protocol is fully implemented.
```
Since tests are now run on C\* 2.2 by default, protocol V4 needs to be "hidden" until everything is implemented.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/LargeDataTest.java:
```diff
@@ -203,9 +205,10 @@ public void wideRows() throws Throwable {
 
     /**
      * Test a batch that writes a row of size 10,000
+     * Since Cassandra 2.2 large batches are rejected, see CASSANDRA-8011.
```
Just a heads up that since CASSANDRA-8011 this test throws InvalidQueryException (batch too large).

tolbertam replied:
I'll update the test so it bumps up against the error threshold, that way it still does the validation it needs to (works with large batches).

adutra commented on driver-core/src/test/java/com/datastax/driver/core/TestUtils.java:
```diff
@@ -442,7 +442,10 @@ public static ProtocolVersion getDesiredProtocolVersion() {
+        } else if (major < 2.2) {
+            return ProtocolVersion.V3;
         } else {
+            // TODO V4 when then entire protocol is implemented
```
Here too we need to inhibit V4 until everything is implemented.

adutra commented on pom.xml:
```diff
@@ -43,7 +43,7 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-    <cassandra.version>2.1.3</cassandra.version>
+    <cassandra.version>2.2.0-beta1</cassandra.version>
```
Default C\* version for tests is now 2.2.0-beta1.

olim7t committed:
Use PK columns in protocol v4 PREPARED response (JAVA-776).

Commit:
Distinguish NULL and UNSET values (JAVA-777).

olim7t committed:
Expose server-side warnings on ExecutionInfo (JAVA-780).

olim7t committed:
Expose new read/write failure exceptions (JAVA-749).

olim7t commented:
The unit test in my commit for JAVA-749 assumes that the default protocol version is V4.

tolbertam commented on driver-core/CHANGELOG.rst:
```diff
@@ -4,9 +4,17 @@ CHANGELOG
 Merged from 2.1 branch:
 
 - [improvement] Unify "Target" enum for schema elements (JAVA-782)
+- [new feature] Expose server-side warnings on ExecutionInfo (JAVA-780)
```
Shouldn't these 2 be under the 2.2.0-rc1 one section and not under "Merged from 2.1 branch"?

olim7t replied:
Yes, that's a merge error, I noticed this while rebasing my udf branch and forgot to report it.

tolbertam commented on driver-core/src/test/java/com/datastax/driver/core/UnsetValuesTest.java:
```diff
@@ -0,0 +1,165 @@
+import static com.datastax.driver.core.ProtocolVersion.V3;
+import static com.datastax.driver.core.ProtocolVersion.V4;
+
+public class UnsetValuesTest {
```
I think these are all duplicates from PreparedStatementTest right?

tolbertam replied:
Looks like it.  I'll delete it.

tolbertam committed:
Integration test improvements for JAVA-783.

- Introduce TestUtils.getDesiredProtocolVersion(maxAllowed) for
  retrieving a ProtocolVersion with the maximum returned being the one
  provided.  Update DynamicConnectionPoolTest and some tests in
  PreparedStatementTest to use it.
- Add WriteFailureException to ExceptionsTest.
- Move test in ReadFailureTest to ExceptionsTest.
- Remove UnsetValuesTest as all of these tests are now present in
  PreparedStatementTest.
- Fix LargeDataTest#testWideRows so it writes up the default max allowed
  batch size in 2.2.0-beta1.

tolbertam commented:
Looks good to me :+1:!  Made a few tweaks to some of the tests and added one for WriteFailureException.  The last commit about fixing should_wait_for_each_contact_point_at_most_once was also backported to 2.1.
